lib/ui/painting/image_descriptor.h

```diff
@@ -119,9 +119,12 @@ class ImageDescriptor : public RefCountedDartWrappable<ImageDescriptor> {
                   const SkImageInfo& image_info,
                   std::optional<size_t> row_bytes);
   ImageDescriptor(sk_sp<SkData> buffer, std::unique_ptr<SkCodec> codec);
+  ImageDescriptor(sk_sp<SkData> buffer,
+                  std::unique_ptr<SkImageGenerator> generator);
 
   sk_sp<SkData> buffer_;
   std::shared_ptr<SkCodecImageGenerator> generator_;
+  std::unique_ptr<SkImageGenerator> platform_image_generator_;
```

Comment:
Instead of having a two fields for the generator with them being conditionally set, it might be more readable to just have a single generator field and in the case where you get a codec, create an `SkCodecImageGenerator` before invoking the designated constructor.

Reply:
The problem is that `SkCodecImageGenerator` has methods on it that `SkImageGenerator` doesn't, and we need them. Specifically:

- `getScaledDimensions`
- `getFrameCount`
- `getRepetitionCount`
- `getFrameInfo`
- `getPixels` with the `SkCodec::Options` parameter

@LeonScroggins - I think I've asked this before, but would it make sense to expose these somehow on SkImageGenerator, perhaps all defaulted to what you'd expect for an image that has 1 frame and cannot be efficiently subscaled?

Alternatively, I could try to refactor this into some more classes.

Reply:
Ah odd. Makes sense. Thanks for the background.

Reply:
> The problem is that `SkCodecImageGenerator` has methods on it that `SkImageGenerator` doesn't, and we need them. Specifically:
> 
> * `getScaledDimensions`
> * `getFrameCount`
> * `getRepetitionCount`
> * `getFrameInfo`
> * `getPixels` with the `SkCodec::Options` parameter
> 
> @LeonScroggins - I think I've asked this before, but would it make sense to expose these somehow on SkImageGenerator, perhaps all defaulted to what you'd expect for an image that has 1 frame and cannot be efficiently subscaled?

Maybe `getScaledDimensions` - we'd need to run it by our API approvers, though. @brianosman mentioned in [this CL](https://skia-review.googlesource.com/c/skia/+/301581) that it doesn't make sense for all `SkImageGenerator`s, though. (That said, I can think of sensible implementations for both of his counterexamples.)

`SkImageGenerator` is expected to create the same image every time - otherwise caches that rely on its `uniqueID()` wouldn't work properly. So the animation-related methods wouldn't make sense on it. But I agree we should have a common API that lets you decode either from the "platform" or the existing `SkCodec` subclasses, whether that's by making a platform version of `SkCodec` or something new.